File: Array/Easy/p3.cpp
// Problem Statement: Given an array of integers, rotating array of elements by k elements either left or right.

// Example 1:
// Input: N = 7, array[] = {1,2,3,4,5,6,7} , k=2 , right
// Output: 6 7 1 2 3 4 5

// Example 2:
// Input: N = 6, array[] = {3,7,8,9,10,11} , k=3 , left
// Output: 9 10 11 3 7 8

#include <iostream>
#include <vector>

using namespace std;

void reverse_array(vector<int> &arr, int start, int end)
{
    while (start < end)
    {
        swap(arr[start], arr[end]);
        start++;
        end--;
    }
}

void left_rotate_array(vector<int> &arr, int n, int k)
{
    reverse_array(arr, 0, k - 1);
    reverse_array(arr, k, n - 1);
    reverse_array(arr, 0, n - 1);
}

void right_rotate_array(vector<int> &arr, int n, int k)
{
    int d = n - k;
    reverse_array(arr, 0, d - 1);
    reverse_array(arr, d, n - 1);
    reverse_array(arr, 0, n - 1);
}

int main(void)
{
    vector<int> arr1 = {1, 2, 3, 4, 5, 6, 7};
    int n1 = arr1.size();

    vector<int> arr2 = {3, 7, 8, 9, 10, 11};
    int n2 = arr2.size();

    left_rotate_array(arr2, n2, 3);  // Time: O(n) Space: O(1)
    right_rotate_array(arr1, n1, 2); // Time: O(n) Space: O(1)

    cout << "Rotated array: \n";
    for (int i = 0; i < n1; i++)
        cout << arr1[i] << " ";
    cout<<endl<<endl;

    cout << "Rotated array: \n";
    for (int i = 0; i < n2; i++)
        cout << arr2[i] << " ";
    cout<<endl;

    return 0;
}

File: Binary_Search/BS_ON_2D_ARRAY/P3.java
// Problem Statement: Given a 0-indexed n x m matrix mat where no two adjacent cells are equal, find any peak element mat[i][j] 
// and return the array [i, j]. A peak element in a 2D grid is an element that is strictly greater than all of its adjacent 
// neighbours to the left, right, top, and bottom. Assume that the entire matrix is surrounded by an outer perimeter with the 
// value -1 in each cell.

// Example:
// Input: mat = [[5, 10, 8], [4, 25, 7], [3, 9, 6]]
// Output: [1, 1]

public class P3 {
    public static int findMaxInColumn(int[][] mat, int n, int m, int col){
        int res = 0, max = mat[0][col];

        for(int i=1; i<n; i++){
            if(mat[i][col] > max){
                res = i;
                max = mat[i][col];
            }
        }

        return res;
    }

    public static int[] findPeakGrid(int[][] mat) {
        int n = mat.length, m = mat[0].length;

        int start = 0, end = m-1;
        while(start<=end){
            int mid = (end-start)/2 + start;

            int elementRow = findMaxInColumn(mat, n, m, mid);
            int left = (mid-1>=0)?mat[elementRow][mid-1]:-1;
            int right = (mid+1<m)?mat[elementRow][mid+1]:-1;

            if(mat[elementRow][mid]>left && mat[elementRow][mid]>right){
                return new int[]{elementRow, mid};
            } else if(left>right){
                end = mid-1;
            } else{
                start = mid+1;
            }
        }

        return new int[]{-1, -1};
    }

    public static void main(String[] args) {
        int[][] matrix = {{5, 10, 8},{4, 25, 7},{3,6,9}};

        int[] res = findPeakGrid(matrix); //Time: O(n * log m) and Space: O(1)
        System.out.println(res[0] + " " + res[1]);
    }
}
